File: sdk/python/trialmcp_client/ledger.py
"""Ledger sub-client for audit chain operations.

Wraps the four Ledger server tools: ``ledger_append``, ``ledger_verify``,
``ledger_query``, and ``ledger_export``.
"""

from __future__ import annotations

import logging
from typing import TYPE_CHECKING, Any

from .config import ServerName
from .exceptions import raise_for_error
from .models import AuditRecord, ChainVerification

if TYPE_CHECKING:
    from .client import TrialMCPClient

logger = logging.getLogger("trialmcp.ledger")

_SERVER = ServerName.LEDGER


class LedgerClient:
    """Sub-client for the ``trialmcp-ledger`` MCP server.

    Provides typed, async wrappers around the four audit ledger tools.
    The ledger maintains a SHA-256 hash-chained, append-only log of all
    MCP tool invocations for 21 CFR Part 11 compliance.

    Parameters
    ----------
    parent:
        The :class:`TrialMCPClient` that owns this sub-client.
    """

    def __init__(self, parent: TrialMCPClient) -> None:
        self._client = parent

    # ------------------------------------------------------------------
    # ledger_append
    # ------------------------------------------------------------------

    async def append(
        self,
        *,
        server: str,
        tool: str,
        caller: str,
        parameters: dict[str, Any],
        result_summary: str,
    ) -> AuditRecord:
        """Append a new record to the audit ledger.

        The server automatically computes the SHA-256 hash and links
        the new record to the previous one.

        Parameters
        ----------
        server:
            MCP server that generated the event.
        tool:
            MCP tool that was invoked.
        caller:
            Pseudonymised actor identifier.
        parameters:
            De-identified tool call parameters.
        result_summary:
            Brief outcome description (must not contain PHI).

        Returns
        -------
        AuditRecord:
            The newly created audit record with its hash.
        """
        params: dict[str, Any] = {
            "server": server,
            "tool": tool,
            "caller": caller,
            "parameters": parameters,
            "result_summary": result_summary,
        }

        result = await self._client.call_tool(_SERVER, "ledger_append", params)
        raise_for_error(result)
        return AuditRecord.from_dict(result)

    # ------------------------------------------------------------------
    # ledger_verify
    # ------------------------------------------------------------------

    async def verify(
        self,
        *,
        start_id: str | None = None,
        end_id: str | None = None,
        full_chain: bool = False,
    ) -> ChainVerification:
        """Verify the integrity of the audit chain.

        Checks that each record's hash correctly links to the previous
        record, detecting any tampering or corruption.

        Parameters
        ----------
        start_id:
            Audit record ID to start verification from.
        end_id:
            Audit record ID to stop verification at.
        full_chain:
            If ``True``, verify the entire chain from genesis.

        Returns
        -------
        ChainVerification:
            Verification result with pass/fail and broken-link details.

        Raises
        ------
        ChainBrokenError:
            If a hash-chain integrity violation is detected.
        """
        params: dict[str, Any] = {}
        if start_id:
            params["start_id"] = start_id
        if end_id:
            params["end_id"] = end_id
        if full_chain:
            params["full_chain"] = True

        result = await self._client.call_tool(_SERVER, "ledger_verify", params)
        raise_for_error(result)

        return ChainVerification(
            valid=result.get("valid", False),
            records_checked=result.get("records_checked", 0),
            first_record_id=result.get("first_record_id"),
            last_record_id=result.get("last_record_id"),
            broken_at_id=result.get("broken_at_id"),
            error_message=result.get("error_message"),
        )

    # ------------------------------------------------------------------
    # ledger_query
    # ------------------------------------------------------------------

    async def query(
        self,
        *,
        server: str | None = None,
        tool: str | None = None,
        caller: str | None = None,
        start_time: str | None = None,
        end_time: str | None = None,
        limit: int = 100,
        offset: int = 0,
    ) -> list[AuditRecord]:
        """Query audit records by various criteria.

        Parameters
        ----------
        server:
            Filter by originating MCP server.
        tool:
            Filter by tool name.
        caller:
            Filter by actor identifier.
        start_time:
            Earliest timestamp (ISO 8601).
        end_time:
            Latest timestamp (ISO 8601).
        limit:
            Maximum number of records to return.
        offset:
            Number of records to skip (for pagination).

        Returns
        -------
        list[AuditRecord]:
            Matching audit records.
        """
        params: dict[str, Any] = {
            "limit": limit,
            "offset": offset,
        }
        if server:
            params["server"] = server
        if tool:
            params["tool"] = tool
        if caller:
            params["caller"] = caller
        if start_time:
            params["start_time"] = start_time
        if end_time:
            params["end_time"] = end_time

        result = await self._client.call_tool(_SERVER, "ledger_query", params)
        raise_for_error(result)

        records_raw = result.get("records", [])
        return [AuditRecord.from_dict(r) for r in records_raw]

    # ------------------------------------------------------------------
    # ledger_export
    # ------------------------------------------------------------------

    async def export(
        self,
        *,
        format: str = "jsonl",
        start_time: str | None = None,
        end_time: str | None = None,
        server: str | None = None,
        include_hashes: bool = True,
    ) -> dict[str, Any]:
        """Export audit records for external compliance review.

        Parameters
        ----------
        format:
            Export format (``"jsonl"``, ``"csv"``, ``"xml"``).
        start_time:
            Earliest timestamp for export window.
        end_time:
            Latest timestamp for export window.
        server:
            Optional filter by originating server.
        include_hashes:
            Whether to include hash chain values in the export.

        Returns
        -------
        dict:
            Export result including the data payload or download reference.
        """
        params: dict[str, Any] = {
            "format": format,
            "include_hashes": include_hashes,
        }
        if start_time:
            params["start_time"] = start_time
        if end_time:
            params["end_time"] = end_time
        if server:
            params["server"] = server

        result = await self._client.call_tool(_SERVER, "ledger_export", params)
        raise_for_error(result)
        return result

    # ------------------------------------------------------------------
    # Convenience helpers
    # ------------------------------------------------------------------

    async def verify_full_chain(self) -> ChainVerification:
        """Verify the entire audit chain from genesis.

        Returns
        -------
        ChainVerification:
            Full chain verification result.
        """
        return await self.verify(full_chain=True)

    async def query_by_caller(
        self,
        caller: str,
        *,
        limit: int = 100,
    ) -> list[AuditRecord]:
        """Query all audit records for a specific actor.

        Parameters
        ----------
        caller:
            The pseudonymised actor identifier.
        limit:
            Maximum records to return.

        Returns
        -------
        list[AuditRecord]:
            Matching audit records.
        """
        return await self.query(caller=caller, limit=limit)

    async def query_by_server(
        self,
        server: str,
        *,
        limit: int = 100,
    ) -> list[AuditRecord]:
        """Query all audit records for a specific MCP server.

        Parameters
        ----------
        server:
            The MCP server name.
        limit:
            Maximum records to return.

        Returns
        -------
        list[AuditRecord]:
            Matching audit records.
        """
        return await self.query(server=server, limit=limit)

    async def get_latest_records(
        self,
        count: int = 10,
    ) -> list[AuditRecord]:
        """Retrieve the most recent audit records.

        Parameters
        ----------
        count:
            Number of recent records to retrieve.

        Returns
        -------
        list[AuditRecord]:
            The most recent audit records.
        """
        return await self.query(limit=count)

File: sdk/typescript/src/ledger.ts
/**
 * Ledger client for the National MCP-PAI Oncology Trials network.
 *
 * Wraps the trialmcp-ledger MCP server tools:
 *   - ledger_append: Append a hash-chained audit record
 *   - ledger_verify: Verify the integrity of the audit chain
 *   - ledger_query: Query audit records with filters
 *   - ledger_export: Export the full audit chain for regulatory review
 *
 * The ledger maintains an immutable, hash-chained sequence of audit
 * records. Each record's hash depends on the previous record's hash,
 * forming a tamper-evident chain per spec/audit.md.
 */

import { TrialMcpClient } from './client';
import { InvalidInputError, ChainBrokenError } from './errors';
import type {
  AuditRecord,
  LedgerAppendParams,
  LedgerVerifyResult,
  LedgerQueryParams,
  LedgerExportResult,
} from './models';

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

const SERVER: 'trialmcp-ledger' = 'trialmcp-ledger';

/** Genesis hash: the previous_hash of the first record in any chain */
const GENESIS_HASH = '0'.repeat(64);

/** SHA-256 hash pattern */
const HASH_PATTERN = /^[0-9a-f]{64}$/;

/** Maximum records per query */
const MAX_QUERY_LIMIT = 1000;

/** MCP servers that produce auditable tool calls */
const AUDITABLE_SERVERS = [
  'trialmcp-authz',
  'trialmcp-fhir',
  'trialmcp-dicom',
  'trialmcp-ledger',
  'trialmcp-provenance',
] as const;

// ---------------------------------------------------------------------------
// Ledger Client
// ---------------------------------------------------------------------------

export class LedgerClient {
  private readonly client: TrialMcpClient;

  constructor(client: TrialMcpClient) {
    this.client = client;
  }

  /**
   * Append a new hash-chained audit record to the ledger.
   *
   * The server computes the SHA-256 hash of the record concatenated
   * with the previous record's hash. The caller provides the operation
   * metadata; the server handles hash chaining.
   *
   * @param params - Audit record fields
   * @returns The appended audit record including its computed hash
   */
  async append(params: LedgerAppendParams): Promise<AuditRecord> {
    this.validateAppendParams(params);

    const result = await this.client.callTool<AuditRecord>(SERVER, 'ledger_append', {
      server: params.server,
      tool: params.tool,
      caller: params.caller,
      result_summary: params.result_summary,
      parameters: params.parameters ?? {},
    });

    return result.data;
  }

  /**
   * Append an audit record for an MCP tool call.
   * Convenience method that constructs the record from tool call metadata.
   */
  async auditToolCall(
    server: string,
    tool: string,
    caller: string,
    resultSummary: string,
    parameters?: Record<string, unknown>,
  ): Promise<AuditRecord> {
    return this.append({
      server,
      tool,
      caller,
      result_summary: resultSummary,
      parameters,
    });
  }

  /**
   * Verify the integrity of the audit chain.
   *
   * Checks that every record's hash correctly chains from the
   * previous record. Detects tampering, insertion, deletion,
   * or reordering of audit records.
   *
   * @returns Verification result with chain length and any issues
   * @throws ChainBrokenError if verification reveals tampering
   */
  async verify(options: { throwOnBroken?: boolean } = {}): Promise<LedgerVerifyResult> {
    const result = await this.client.callTool<LedgerVerifyResult>(
      SERVER,
      'ledger_verify',
      {},
    );

    const verification = result.data;

    if (!verification.valid && options.throwOnBroken) {
      throw new ChainBrokenError(
        `Audit chain integrity verification failed: ${verification.reason ?? 'unknown reason'}`,
        {
          server: SERVER,
          tool: 'ledger_verify',
          details: {
            chain_length: verification.length,
            broken_at_index: verification.broken_at_index,
          } as Record<string, unknown>,
        },
      );
    }

    return verification;
  }

  /**
   * Assert that the audit chain is valid. Throws on broken chain.
   */
  async assertValid(): Promise<LedgerVerifyResult> {
    return this.verify({ throwOnBroken: true });
  }

  /**
   * Query audit records with optional filters.
   *
   * Supports filtering by server, tool, caller, and time range.
   * Results are returned in chronological order.
   *
   * @param params - Query filters
   * @returns Matching audit records
   */
  async query(params: LedgerQueryParams = {}): Promise<AuditRecord[]> {
    this.validateQueryParams(params);

    const toolParams: Record<string, unknown> = {};

    if (params.server) toolParams.server = params.server;
    if (params.tool) toolParams.tool = params.tool;
    if (params.caller) toolParams.caller = params.caller;
    if (params.from_timestamp) toolParams.from_timestamp = params.from_timestamp;
    if (params.to_timestamp) toolParams.to_timestamp = params.to_timestamp;
    if (params.limit !== undefined) {
      toolParams.limit = Math.min(params.limit, MAX_QUERY_LIMIT);
    }

    const result = await this.client.callTool<AuditRecord[]>(
      SERVER,
      'ledger_query',
      toolParams,
    );

    return result.data;
  }

  /**
   * Query records for a specific server.
   */
  async queryByServer(server: string, limit?: number): Promise<AuditRecord[]> {
    return this.query({ server, limit });
  }

  /**
   * Query records for a specific tool.
   */
  async queryByTool(tool: string, limit?: number): Promise<AuditRecord[]> {
    return this.query({ tool, limit });
  }

  /**
   * Query records for a specific caller.
   */
  async queryByCaller(caller: string, limit?: number): Promise<AuditRecord[]> {
    return this.query({ caller, limit });
  }

  /**
   * Query records within a time range.
   */
  async queryByTimeRange(
    from: string,
    to: string,
    limit?: number,
  ): Promise<AuditRecord[]> {
    return this.query({ from_timestamp: from, to_timestamp: to, limit });
  }

  /**
   * Export the complete audit chain for regulatory review.
   *
   * Returns all audit records in chain order with a checksum
   * of the complete export for integrity verification.
   *
   * @returns Full audit chain export with metadata
   */
  async export(): Promise<LedgerExportResult> {
    const result = await this.client.callTool<LedgerExportResult>(
      SERVER,
      'ledger_export',
      {},
    );

    return result.data;
  }

  /**
   * Count audit records matching the given filters.
   */
  async count(params: LedgerQueryParams = {}): Promise<number> {
    const records = await this.query({ ...params, limit: MAX_QUERY_LIMIT });
    return records.length;
  }

  /**
   * Get the latest N audit records.
   */
  async latest(count: number = 10): Promise<AuditRecord[]> {
    const records = await this.query({ limit: count });
    return records.slice(-count);
  }

  // -------------------------------------------------------------------------
  // Validation
  // -------------------------------------------------------------------------

  private validateAppendParams(params: LedgerAppendParams): void {
    if (!params.server || params.server.trim().length === 0) {
      throw new InvalidInputError('server is required for ledger append', {
        server: SERVER,
        tool: 'ledger_append',
        details: { field: 'server' },
      });
    }

    if (!params.tool || params.tool.trim().length === 0) {
      throw new InvalidInputError('tool is required for ledger append', {
        server: SERVER,
        tool: 'ledger_append',
        details: { field: 'tool' },
      });
    }

    if (!params.caller || params.caller.trim().length === 0) {
      throw new InvalidInputError('caller is required for ledger append', {
        server: SERVER,
        tool: 'ledger_append',
        details: { field: 'caller' },
      });
    }

    if (!params.result_summary || params.result_summary.trim().length === 0) {
      throw new InvalidInputError('result_summary is required for ledger append', {
        server: SERVER,
        tool: 'ledger_append',
        details: { field: 'result_summary' },
      });
    }
  }

  private validateQueryParams(params: LedgerQueryParams): void {
    if (params.limit !== undefined && (params.limit < 1 || params.limit > MAX_QUERY_LIMIT)) {
      throw new InvalidInputError(
        `Query limit must be between 1 and ${MAX_QUERY_LIMIT}, got ${params.limit}`,
        { server: SERVER, tool: 'ledger_query', details: { field: 'limit' } },
      );
    }

    if (params.from_timestamp) {
      const date = new Date(params.from_timestamp);
      if (isNaN(date.getTime())) {
        throw new InvalidInputError(
          `Invalid from_timestamp: '${params.from_timestamp}'. Must be ISO 8601`,
          { server: SERVER, tool: 'ledger_query', details: { field: 'from_timestamp' } },
        );
      }
    }

    if (params.to_timestamp) {
      const date = new Date(params.to_timestamp);
      if (isNaN(date.getTime())) {
        throw new InvalidInputError(
          `Invalid to_timestamp: '${params.to_timestamp}'. Must be ISO 8601`,
          { server: SERVER, tool: 'ledger_query', details: { field: 'to_timestamp' } },
        );
      }
    }
  }
}
